cli/internal/run/global_hash.go

```diff
@@ -47,28 +47,40 @@ type oldGlobalHashable struct {
 }
 
 // calculateGlobalHashFromHashable returns a hash string from the globalHashable
-func calculateGlobalHashFromHashable(named GlobalHashable) (string, error) {
-	// When we aren't in infer mode, we can hash the whole object
-	if named.envMode != util.Infer {
-		return fs.HashObject(named)
-	}
+func calculateGlobalHashFromHashable(full GlobalHashable) (string, error) {
+	switch full.envMode {
+	case util.Infer:
+		if full.envVarPassthroughs != nil {
+			// In infer mode, if there is any passThru config (even if it is an empty array)
+			// we'll hash the whole object, so we can detect changes to that config
+			// Further, resolve the envMode to the concrete value.
+			full.envMode = util.Strict
+			return fs.HashObject(full)
+		}
 
-	// In infer mode, if there is any passThru config (even if it is an empty array)
-	// we'll hash the whole object, so we can detect changes to that config
-	if named.envVarPassthroughs != nil {
-		return fs.HashObject(named)
+		// If we're in infer mode, and there is no global pass through config,
+		// we use the old struct layout. this will be true for everyone not using the strict env
+		// feature, and we don't want to break their cache.
+		return fs.HashObject(oldGlobalHashable{
+			globalFileHashMap:    full.globalFileHashMap,
+			rootExternalDepsHash: full.rootExternalDepsHash,
+			envVars:              full.envVars.All.ToHashable(),
+			globalCacheKey:       full.globalCacheKey,
+			pipeline:             full.pipeline,
+		})
+	case util.Loose:
+		// Remove the passthroughs from hash consideration if we're explicitly loose.
+		full.envVarPassthroughs = nil
+		return fs.HashObject(full)
+	case util.Strict:
+		// Collapse `nil` and `[]` in strict mode.
+		if full.envVarPassthroughs == nil {
+			full.envVarPassthroughs = make([]string, 0)
+		}
+		return fs.HashObject(full)
+	default:
+		panic("unimplemented environment mode")
```

Comment:
Do we have lints set up to statically infer exhaustiveness (https://pkg.go.dev/github.com/nishanths/exhaustive@v0.10.0)? I don't particularly like this but if we don't have it it's probably not worth setting up now.

Reply:
We don't have exhaustiveness lints. I looked for it, it exists, and this is the only enum in the codebase that it would check.

I yearn for Rust's approach here.